src/Editor/ImGui/ImGuizmo.cpp

```diff
@@ -0,0 +1,1719 @@
+                gContext.mBoundsBestAxis = bestAxis;
+                int indices[] = { secondAxis , thirdAxis };
+                gContext.mBoundsAxis[0] = indices[i % 2];
+                gContext.mBoundsAxis[1] = -1;
+
+                gContext.mBoundsLocalPivot.Set(0.f);
+                gContext.mBoundsLocalPivot[gContext.mBoundsAxis[0]] = aabb[oppositeIndex][indices[i % 2]];// bounds[gContext.mBoundsAxis[0]] * (((i + 1) & 2) ? 1.f : -1.f);
+
+                gContext.mbUsingBounds = true;
+                gContext.mBoundsMatrix = gContext.mModelSource;
+            }
+        }
+
+        if (gContext.mbUsingBounds)
+        {
+            matrix_t scale;
+            scale.SetToIdentity();
+
+            // compute projected mouse position on plan
+            const float len = IntersectRayPlane(gContext.mRayOrigin, gContext.mRayVector, gContext.mBoundsPlan);
+            vec_t newPos = gContext.mRayOrigin + gContext.mRayVector * len;
+
+            // compute a reference and delta vectors base on mouse move
+            vec_t deltaVector = (newPos - gContext.mBoundsPivot).Abs();
+            vec_t referenceVector = (gContext.mBoundsAnchor - gContext.mBoundsPivot).Abs();
+
+            // for 1 or 2 axes, compute a ratio that's used for scale and snap it based on resulting length
+            for (int i = 0; i < 2; i++)
+            {
+                int axisIndex = gContext.mBoundsAxis[i];
+                if (axisIndex == -1)
+                    continue;
+
+                float ratioAxis = 1.f;
+                vec_t axisDir = gContext.mBoundsMatrix.component[axisIndex].Abs();
+
+                float dtAxis = axisDir.Dot(referenceVector);
+                float boundSize = bounds[axisIndex + 3] - bounds[axisIndex];
+                if (dtAxis > FLT_EPSILON)
+                    ratioAxis = axisDir.Dot(deltaVector) / dtAxis;
+
+                if (snapValues)
+                {
+                    float length = boundSize * ratioAxis;
+                    ComputeSnap(&length, snapValues[axisIndex]);
+                    if (boundSize > FLT_EPSILON)
+                        ratioAxis = length / boundSize;
+                }
+                scale.component[axisIndex] *= ratioAxis;
+            }
+
+            // transform matrix
+            matrix_t preScale, postScale;
+            preScale.Translation(-gContext.mBoundsLocalPivot);
+            postScale.Translation(gContext.mBoundsLocalPivot);
+            matrix_t res = preScale * scale * postScale * gContext.mBoundsMatrix;
+            *matrix = res;
+
+            // info text
+            char tmps[512];
+            ImVec2 destinationPosOnScreen = worldToPos(gContext.mModel.v.position, gContext.mViewProjection);
+            ImFormatString(tmps, sizeof(tmps), "X: %.2f Y: %.2f Z:%.2f"
+                , (bounds[3] - bounds[0]) * gContext.mBoundsMatrix.component[0].Length() * scale.component[0].Length()
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') comma should not be preceded by whitespace [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AT009)